cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/DefaultProvisionerConfigProvider.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.provision;
+
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.utils.DirUtils;
+import com.google.common.base.Charsets;
+import com.google.common.base.Splitter;
+import com.google.common.collect.ImmutableList;
+import com.google.common.io.Files;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonArray;
+import com.google.gson.stream.JsonReader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.Reader;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.inject.Inject;
+
+/**
+ * Default implementation of the ProvsionerConfigProvider. It expects a json file for from each module dir and
+ * expects a "configuration-groups" in the json file
+ */
+public class DefaultProvisionerConfigProvider implements ProvisionerConfigProvider {
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultProvisionerConfigProvider.class);
+  private static final Gson GSON = new GsonBuilder().create();
+
+  private List<String> extDirs;
+
+  @Inject
+  public DefaultProvisionerConfigProvider(CConfiguration cConf) {
+    String extDirectory = cConf.get(Constants.Provisioner.EXTENSIONS_DIR);
+    this.extDirs = ImmutableList.copyOf(Splitter.on(';').omitEmptyStrings().trimResults().split(extDirectory));
+  }
+
+  @Override
+  public Map<String, ProvisionerConfig> loadProvisionerConfigs() {
+    Map<String, ProvisionerConfig> results = new HashMap<>();
+
+    for (String dir : extDirs) {
+      File extDir = new File(dir);
+      if (!extDir.isDirectory()) {
+        continue;
+      }
+
+      List<File> files = new ArrayList<>(DirUtils.listFiles(extDir));
+      Collections.sort(files);
+      for (File moduleDir : files) {
+        if (!moduleDir.isDirectory()) {
+          continue;
+        }
+
+        List<File> jsonFiles =
+          DirUtils.listFiles(moduleDir,
+                             file -> !file.isDirectory() && file.getName().equals(moduleDir.getName() + ".json"));
```

Comment:
the module directory is not guaranteed to be the provisioner name. In fact, you can have multiple provisioners in the same directory. For example, gce-dataproc directory could have dataproc-singlenode, dataproc-multinode provisioners in them. I think it's better for this method to take in a set of provisioner names that you need to find configs for.